File: shoope/31-next-permutation/31.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int i = nums.size() - 2;
        while (i >= 0 && nums[i] >= nums[i+1]) {
            --i;
        }
        if (i >= 0) {
            int j = nums.size() - 1;
            while (i < j && nums[i] >= nums[j]) {
                --j;
            }
            swap(nums[i], nums[j]);
        }

        reverse(nums.begin() + i + 1, nums.end());
    }
};

int main(int argc, char** args) {
    Solution sol = Solution();
	vector<int> nums = {1,2,3};
	sol.nextPermutation(nums);
    for (auto a : nums) {
        cout << a << " ";
    }
    cout << endl;

    nums = {3,2,1};
	sol.nextPermutation(nums);
    for (auto a : nums) {
        cout << a << " ";
    }
    cout << endl;

	nums = {1,1,5};
	sol.nextPermutation(nums);
    for (auto a : nums) {
        cout << a << " ";
    }
    cout << endl;
    
    nums = {1};
	sol.nextPermutation(nums);
    for (auto a : nums) {
        cout << a << " ";
    }
    cout << endl;

    nums = {1,5,1};
	sol.nextPermutation(nums);
    for (auto a : nums) {
        cout << a << " ";
    }
    cout << endl;

    nums = {5,1,1};
	sol.nextPermutation(nums);
    for (auto a : nums) {
        cout << a << " ";
    }
    cout << endl;

    return 0;
}

File: shoope/31-next-permutation/31.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

import sys
from typing import List


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i = len(nums) - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
        if i >= 0:
            j = len(nums) - 1
            while i < j and nums[i] >= nums[j]:
                j -= 1
            nums[i], nums[j] = nums[j], nums[i]

        j, i = len(nums) - 1, i + 1
        while i < j:
            nums[i], nums[j] = nums[j], nums[i]
            i += 1
            j -= 1


if __name__ == '__main__':
    sol = Solution()
    nums = [1, 2, 3]
    sol.nextPermutation(nums)
    print(nums)
    nums = [3, 2, 1]
    sol.nextPermutation(nums)
    print(nums)
    nums = [1, 1, 5]
    sol.nextPermutation(nums)
    print(nums)
    nums = [1]
    sol.nextPermutation(nums)
    print(nums)
    nums = [1, 5, 1]
    sol.nextPermutation(nums)
    print(nums)
    nums = [5, 1, 1]
    sol.nextPermutation(nums)
    print(nums)
